libraries/GCS_MAVLink/GCS.h

```diff
@@ -446,11 +440,85 @@ class GCS_MAVLINK
     char _perf_packet_name[16];
     char _perf_update_name[16];
 
-    // deferred message handling.  We size the deferred_message
-    // ringbuffer so we can defer every message type
-    enum ap_message deferred_messages[MSG_LAST];
-    uint8_t next_deferred_message;
-    uint8_t num_deferred_messages;
+    // outbound ("deferred message") queue.
+
+    // this is a simple array of deferred_message_t.  We keep a
+    // pointer to the next message that is to be sent (by offset,
+    // next_deferred_message_to_send); we need only check its
+    // send_not_before_ms entry to determine if we have work to do.
+    // In the event the message is sent, its send_not_before_ms is
+    // updated, and we linearly scan the array to find the next
+    // message that should be sent.  We may immediately send that
+    // message, or may find it isn't due to be sent yet, or that we
+    // have hit some sort of time limit.  An interval of 0 indicates
+    // that this message should not be rescheduled once sent.  An
+    // send_not_before_ms of zero indicates this message should not be
+    // sent.
+
+    // structure holding information about a scheduled message.
+    struct deferred_message_t {
+        enum ap_message id;
```

Comment:
we could removed the id field, and use ptr-difference to get the id, then we can make send_not_before_ms to be uint16_t. That will reduce the memory overhead of this system from 1600 bytes to 800 bytes.
It would also mean that handling time wrap will be tested every 65s, so we'll be confident it will work
we'd need a millis16() function to suppot it